source/ui/dataman/AMScanViewPlotSelectedToolsView.cpp

```diff
@@ -0,0 +1,130 @@
+#include "AMScanViewPlotSelectedToolsView.h"
+#include "ui/dataman/AMScanViewPlotToolView.h"
+#include "MPlot/MPlotTools.h"
+
+AMScanViewPlotSelectedToolsView::AMScanViewPlotSelectedToolsView(AMScanViewPlotTools *tools, QWidget *parent) :
+    QWidget(parent)
+{
+	// Initialize member variables.
+
+	tools_ = 0;
+
+	// Create and set layouts.
+
+	layout_ = new QHBoxLayout();
+	layout_->setMargin(0);
+
+	setLayout(layout_);
+
+	// Current settings.
+
+	setTools(tools);
+}
+
+AMScanViewPlotSelectedToolsView::~AMScanViewPlotSelectedToolsView()
+{
+
+}
+
+void AMScanViewPlotSelectedToolsView::setTools(AMScanViewPlotTools *newTools)
+{
+	if (tools_ != newTools) {
+
+		if (tools_) {
+			disconnect( tools_, 0, this, 0 );
+		}
+
+		tools_ = newTools;
+
+		if (tools_) {
+			connect( tools_, SIGNAL(toolsChanged(QList<MPlotAbstractTool*>)), this, SLOT(refresh()) );
+			connect( tools_, SIGNAL(selectedToolsChanged(QList<MPlotAbstractTool*>)), this, SLOT(update()) );
+		}
+
+		refresh();
+
+		emit toolsChanged(tools_);
+	}
+}
+
+void AMScanViewPlotSelectedToolsView::clear()
+{
+	foreach (QWidget *view, toolViews_) {
+		toolViews_.removeOne(view);
+		layout_->removeWidget(view);
+		view->deleteLater();
+	}
+}
+
+void AMScanViewPlotSelectedToolsView::update()
+{
+	if (tools_) {
+
+		// Iterate through all tool views and show the views for selected tools.
+		// Hide the views of unselected tools.
+
+		int viewCount = toolViews_.count();
+
+		for (int viewIndex = 0; viewIndex < viewCount; viewIndex++) {
+			if (viewIndex < tools_->tools().count()) {
+				MPlotAbstractTool *tool = tools_->tools().at(viewIndex);
+				QWidget *toolView = toolViews_.at(viewIndex);
+
+				if (tool && toolView) {
+					if (tools_->isSelectedTool(tool))
+						toolView->show();
+					else
+						toolView->hide();
+				}
+			}
+		}
+	}
+}
+
+void AMScanViewPlotSelectedToolsView::refresh()
+{
+	// Clear the view.
+
+	clear();
+
+	// If we have valid tools, create the view.
+
+	if (tools_) {
+
+		// Create tool view for each tool and add to list of tool views.
+
+		int toolCount = tools_->tools().count();
+
+		for (int toolIndex = 0; toolIndex < toolCount; toolIndex++) {
+			MPlotAbstractTool *tool = tools_->tools().at(toolIndex);
+
+			if (tool) {
+				QWidget *toolView = createToolView(tool);
+				toolViews_ << toolView;
+				layout_->addWidget(toolView);
```

Comment:
The create tool view seems to have some cases where it can return null, so this here seems to be the source of some "Cannot add null widget to QHBoxView" debug messages. I believe the createToolView needs to either be extended to cover all cases, or if this isn't the intended behaviour this block needs to check if the returned toolView is null before adding it to the layout.